### PR TITLE
PM-14995 Hide TOTP for non premium org items even if individual user has premium account

### PR DESCRIPTION
**Pull request by dseverns-livefront**
## 🎟️ Tracking
https://bitwarden.atlassian.net/browse/PM-14995

## 📔 Objective
- A user with premium should not be able to use TOTP codes on items held by a non premium organization.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/data/autofill/manager/AutofillTotpManagerImpl.kt [hunk 1]
```diff
@@ -24,8 +25,15 @@ class AutofillTotpManagerImpl(
 ) : AutofillTotpManager {
     override suspend fun tryCopyTotpToClipboard(cipherView: CipherView) {
         if (settingsRepository.isAutoCopyTotpDisabled) return
+        val organizationPremiumStatusMap = authRepository
+            .userStateFlow
+            .value
+            ?.activeAccount
+            ?.getOrganizationPremiumStatusMap()
+            ?: emptyMap()
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/autofill/manager/AutofillTotpManagerImpl.kt:33]: Can we just do `orEmpty()`

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/vault/util/VaultDataExtensions.kt [hunk 2]
```diff
@@ -73,14 +74,24 @@ fun VaultData.toViewState(
...
+        }
+        val userOwnedTotpItems = totpItemsGroupedByOwnership[false]
+            ?.filter {
+                it.login?.totp != null && isPremium
+            }
+            ?: emptyList()
+        val organizationOwnedTotpItems = totpItemsGroupedByOwnership[true]
+            ?.filter {
+            it.login?.totp != null &&
+                (organizationPremiumStatusMap[it.id] == true || it.organizationUseTotp)
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/vault/util/VaultDataExtensions.kt:88]: Formatting here

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/vault/util/VaultDataExtensions.kt [hunk 3]
```diff
@@ -73,14 +74,24 @@ fun VaultData.toViewState(
...
+            }
+            ?: emptyList()
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/vault/util/VaultDataExtensions.kt:90]: Can we do more `orEmpty()`s

File: app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/verificationcode/VerificationCodeViewModel.kt [hunk 4]
```diff
@@ -319,8 +320,11 @@ class VerificationCodeViewModel @Inject constructor(
         authCodes: List<VerificationCodeItem>,
         userAccount: UserState.Account?,
     ): DataState<List<VerificationCodeItem>> {
+        val orgPremiumStatusMap = userAccount?.getOrganizationPremiumStatusMap() ?: emptyMap()
```
Comment (SaintPatrck) [inline app/src/main/java/com/x8bit/bitwarden/ui/vault/feature/verificationcode/VerificationCodeViewModel.kt]: Another `orEmpty` candidate.

**Review by SaintPatrck — APPROVED** [review]

**Comment by bsteimel** [comment]
This feature is in direct violation to the outline of your user agreements and feature list.  

"Storage of keys is available to all accounts, however TOTP code generation requires Premium or membership to a paid organization (Families, Teams, or Enterprise)."  https://bitwarden.com/help/integrated-authenticator/

OR not AND

**Comment by JaggedJax** [comment]
@dseverns-livefront @SaintPatrck  this PR goes against all of the copy on the Bitwarden website[1][2] and removes access to TOTP codes for paid customers with ZERO notice. This needs to be reverted.

image

[1]https://bitwarden.com/help/about-organizations/
[2]https://bitwarden.com/help/integrated-authenticator/

**Pull request merged**